chain/gen/genesis/fevm.go

```diff
@@ -0,0 +1,113 @@
+package genesis
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	actorstypes "github.com/filecoin-project/go-state-types/actors"
+	"github.com/filecoin-project/go-state-types/big"
+	builtintypes "github.com/filecoin-project/go-state-types/builtin"
+	evm10 "github.com/filecoin-project/go-state-types/builtin/v10/evm"
+	init10 "github.com/filecoin-project/go-state-types/builtin/v10/init"
+	"github.com/filecoin-project/go-state-types/manifest"
+	"github.com/filecoin-project/go-state-types/network"
+
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/consensus/filcns"
+	"github.com/filecoin-project/lotus/chain/state"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/vm"
+)
+
+func SetupFEVM(ctx context.Context, cs *store.ChainStore, sys vm.SyscallBuilder, sroot cid.Cid, nv network.Version) (cid.Cid, error) {
+	av, err := actorstypes.VersionForNetwork(nv)
+	if err != nil {
+		return cid.Undef, fmt.Errorf("failed to get actors version for network version %d: %w", nv, err)
+	}
+
+	if av < actorstypes.Version10 {
+		// Not defined before version 10; migration has to setup.
+		return sroot, nil
+	}
+
+	csc := func(context.Context, abi.ChainEpoch, *state.StateTree) (abi.TokenAmount, error) {
+		return big.Zero(), nil
+	}
+
+	newVM := func(base cid.Cid) (vm.Interface, error) {
+		vmopt := &vm.VMOpts{
+			StateBase:      base,
+			Epoch:          0,
+			Rand:           &fakeRand{},
+			Bstore:         cs.StateBlockstore(),
+			Actors:         filcns.NewActorRegistry(),
+			Syscalls:       mkFakedSigSyscalls(sys),
+			CircSupplyCalc: csc,
+			NetworkVersion: nv,
+			BaseFee:        big.Zero(),
+		}
+
+		return vm.NewVM(ctx, vmopt)
+	}
+
+	genesisVm, err := newVM(sroot)
+	if err != nil {
+		return cid.Undef, fmt.Errorf("creating vm: %w", err)
+	}
+
+	// The ETH0 address is occupied by an empty contract EVM actor
+	evmCodeCid, ok := actors.GetActorCodeID(av, manifest.EvmKey)
+	if !ok {
+		return cid.Undef, fmt.Errorf("failed to get CodeCID for EVM during genesis")
+	}
+
+	// initcode:
+	// %push(code_end - code_begin)
+	// dup1
+	// %push(code_begin)
+	// push1 0x00
+	// codecopy
+	// push1 0x00
+	// return
+	// code_begin:
+	// push1 0x00
+	// push1 0x00
+	// return
+	// code_end:
+	initcode, err := hex.DecodeString("600580600b6000396000f360006000f3")
+	if err != nil {
+		return cid.Undef, fmt.Errorf("failed to parse ETH0 init code during genesis: %w", err)
+	}
+
+	ctorParams := &evm10.ConstructorParams{
+		Creator: make([]byte, 20), // self!
+		// TODO we have a bunch of bugs in the evm constructor around empty contracts
+		// - empty init code is not allowed
+		// - returning an empty contract is not allowed
+		// So this uses code that constructs a just return contract until that can be fixed
+		// and we can pass an empty byte array
```

Comment:
Have those bugs been fixed since this todo was written?

Reply:
Not able to find any of these TODOs in the latest version of the code. Were you reviewing an old commit? (Maybe I could find out by clicking thirough the UI, but the UI is really sluggish on such a big PR) What review strategy are you using here to dig up old TODOs?